### ISSUE TITLE
[feature] Redirect on external TMS server for out-of-range tiles

### ISSUE THREAD

**Issue by nono303** [telemaxx/mapsforgesrv#29]
Hi @telemaxx, @JFritzle
Another quick feature in patch mode for easier intergration in both java8 & java11 branch: 
`Url pattern [ex. https://a.tile.openstreetmap.fr/osmfr/{z}/{x}/{y}.png] of an external TMS server used in redirect for out-of-range tiles (default: unset - disabled)`

```
diff --git a/README.md b/README.md
index f40d2cc..ae889d2 100644
--- a/README.md
+++ b/README.md
@@ -79,8 +79,9 @@ Command parameters:
 	21. -mit [min-thread]			Minimum pool size for rendering job (default: 0)

 	22. -idl [idle-timeout]			Maximum thread idle time in milliseconds (default: 0 - disabled)

 	23. -ct  [connectors]			Comma-separated list of enabled server connector protocol(s) [http11,proxy,h2c] (default: http11)

-	24. -c   [config]			Config file overriding cmd line parameters (default: none)

-	25. -h   [help]				Print the help text and exit 

+	24. -tms [outofrange_tms]		Url pattern [ex. https://a.tile.openstreetmap.fr/osmfr/{z}/{x}/{y}.png] of an external TMS server used in redirect for out-of-range tiles (default: unset - disabled)

+	25. -c   [config]			Config file overriding cmd line parameters (default: none)

+	26. -h   [help]				Print the help text and exit 

 

 

 Example:

diff --git a/mapsforgesrv/config.sample.properties b/mapsforgesrv/config.sample.properties
index c8346bd..4aa98c0 100644
--- a/mapsforgesrv/config.sample.properties
+++ b/mapsforgesrv/config.sample.properties
@@ -27,4 +27,5 @@ cache-control=2592000
 device-scale=1

 user-scale=1

 text-scale=1

-symbol-scale=1
\ No newline at end of file
+symbol-scale=1

+outofrange_tms=https://a.tile.openstreetmap.fr/osmfr/{z}/{x}/{y}.png
\ No newline at end of file
diff --git a/mapsforgesrv/src/main/java/com/telemaxx/mapsforgesrv/MapsforgeConfig.java b/mapsforgesrv/src/main/java/com/telemaxx/mapsforgesrv/MapsforgeConfig.java
index f627dab..5dd77fa 100644
--- a/mapsforgesrv/src/main/java/com/telemaxx/mapsforgesrv/MapsforgeConfig.java
+++ b/mapsforgesrv/src/main/java/com/telemaxx/mapsforgesrv/MapsforgeConfig.java
@@ -54,6 +54,7 @@ public class MapsforgeConfig {
 	private float userScale;

 	private float textScale;

 	private float symbolScale;

+	private String outOfRangeTms = null;

 

 	private final static long DEFAULTCACHECONTROL = 0;

 	private final static int DEFAULTSERVERPORT = 8080;

@@ -232,6 +233,11 @@ public class MapsforgeConfig {
 				.longOpt("connectors") //$NON-NLS-1$

 				.desc("Comma-separated list of enabled server connector protocol(s) [http11,proxy,h2c] (default: http11)") //$NON-NLS-1$

 				.required(false).hasArg(true).build());

+		

+		options.addOption(Option.builder("tms") //$NON-NLS-1$

+				.longOpt("outofrange_tms") //$NON-NLS-1$

+				.desc("Url pattern [ex. https://a.tile.openstreetmap.fr/osmfr/{z}/{x}/{y}.png] of an external TMS server used in redirect for out-of-range tiles (default: unset - disabled)") //$NON-NLS-1$

+				.required(false).hasArg(true).build());

 

 		options.addOption(Option.builder("c") //$NON-NLS-1$

 				.longOpt("config") //$NON-NLS-1$

@@ -535,6 +541,7 @@ public class MapsforgeConfig {
 		textScale = (float) parseNumber(DEFAULTTEXTSCALE, "text-scale", 0., null, "Text scale factor",true); //$NON-NLS-1$ //$NON-NLS-2$

 		symbolScale = (float) parseNumber(DEFAULTSYMBOLSCALE, "symbol-scale", 0., null, "Symbol scale factor",true); //$NON-NLS-1$ //$NON-NLS-2$

 		cacheControl = (long) parseNumber(DEFAULTCACHECONTROL, "cache-control", 0, null, "Browser cache ttl",false); //$NON-NLS-1$ //$NON-NLS-2$

+		outOfRangeTms = parseString(null, "outofrange_tms", null, "Out of range TMS url"); //$NON-NLS-1$ //$NON-NLS-2$

 		maxQueueSize = (int) parseNumber(DEFAULTSERVERMAXQUEUESIZE, "max-queuesize", 0, null, "Server max queue size",false); //$NON-NLS-1$ //$NON-NLS-2$

 		maxThreads = (int) parseNumber(DEFAULTSERVERMAXTHREADS, "max-thread", 1, null, "Server max thread(s)",false); //$NON-NLS-1$ //$NON-NLS-2$

 		minThreads = (int) parseNumber(DEFAULTSERVERMINTHREADS, "min-thread", 0, null, "Server min thread(s)",false); //$NON-NLS-1$ //$NON-NLS-2$

@@ -629,6 +636,10 @@ public class MapsforgeConfig {
 	public String[] getServerConnectors() {

 		return this.serverConnectors;

 	}

+	

+	public String getOutOfRangeTms() {

+		return this.outOfRangeTms;

+	}

 

 	public double getGammaValue() {

 		return this.gammaValue;

diff --git a/mapsforgesrv/src/main/java/com/telemaxx/mapsforgesrv/MapsforgeHandler.java b/mapsforgesrv/src/main/java/com/telemaxx/mapsforgesrv/MapsforgeHandler.java
index e488ae7..bab2a9d 100644
--- a/mapsforgesrv/src/main/java/com/telemaxx/mapsforgesrv/MapsforgeHandler.java
+++ b/mapsforgesrv/src/main/java/com/telemaxx/mapsforgesrv/MapsforgeHandler.java
@@ -105,6 +105,8 @@ public class MapsforgeHandler extends AbstractHandler {
 	protected final File themeFile;
 	protected final String themeFileStyle;
 	
+	protected final String outOfRangeTms;
+	
 	protected int blackValue;
 	protected double gammaValue;
 	
@@ -136,6 +138,8 @@ public class MapsforgeHandler extends AbstractHandler {
 		this.themeFile = mapsforgeConfig.getThemeFile();
 		this.themeFileStyle = mapsforgeConfig.getThemeFileStyle();
 		
+		this.outOfRangeTms = mapsforgeConfig.getOutOfRangeTms();
+		
 		this.blackValue = mapsforgeConfig.getBlackValue();
 		this.gammaValue = mapsforgeConfig.getGammaValue();
 		// first apply gamma correction and then contrast-stretching
@@ -503,9 +507,16 @@ public class MapsforgeHandler extends AbstractHandler {
 				}
 				response.setStatus(200);			
 			} else {
-				// TODO temp
-				image = BI_NOCONTENT;
-				response.setStatus(404);
+				if(this.outOfRangeTms != null) {
+					response.setStatus(HttpServletResponse.SC_MOVED_TEMPORARILY);
+					String redirecturl = this.outOfRangeTms.replace("{x}", x+"").replace("{y}", y+"").replace("{z}", z+"");
+					response.setHeader("Location", redirecturl);
+					logger.info("out-of-range redirect '"+redirecturl+"'");
+					return;
+				} else {
+					image = BI_NOCONTENT;
+					response.setStatus(404);
+				}
 			}
 			if (mapsforgeConfig.getCacheControl() > 0) {
 				response.addHeader("Cache-Control", "public, max-age=" + mapsforgeConfig.getCacheControl()); //$NON-NLS-1$ //$NON-NLS-2$

```

**Comment by telemaxx** [comment]
when i understand correct idea is: when i am outside of my map file(s) then use external tms server.
correct? when yes, that idea is good, sometimes my program i somewhere and i see only a white map. then i click "zoom to track" in mytourbook or i display hillshading to have an overview.

**Comment by nono303** [comment]
you're correct ;)

**Comment by nono303** [comment]
hmmm... 
What happen when you hit `http://127.0.0.1:8080/8/144/80.png` ?

I tried with same config as yours (germany north)
`-tms https://a.tile.openstreetmap.fr/osmfr/{z}/{x}/{y}.png` in cli or `outofrange_tms=https://a.tile.openstreetmap.fr/osmfr/{z}/{x}/{y}.png` in config file **both** with good result:

```
2022-03-14 22:29:00.865 main@MapsforgeConfig:525 - ################## CONFIG INIT ##################
2022-03-14 22:29:00.870 main@MapsforgeConfig:336 - Listening TCP port     : defined [8484]
2022-03-14 22:29:00.870 main@MapsforgeConfig:358 - Server interface       : defined [all]
2022-03-14 22:29:00.873 main@MapsforgeConfig:463 - Map file(s)            : defined [{F:\maps\oam\Germany-North_oam.osm.map}]
2022-03-14 22:29:00.874 main@MapsforgeConfig:388 - Theme file             : defined [F:\maps\oam\themes\Elevate_5.0.0\Elevate.xml]
2022-03-14 22:29:00.874 main@MapsforgeConfig:358 - Theme style            : defined [elv-hiking]
2022-03-14 22:29:00.881 main@MapsforgeConfig:477 - Theme overlay(s)       : defined [{elv-h_routes,elv-h_s_routes,elv-waymarks,elv-road_surfaces-h,elv-acc_allowed-h,elv-settlements-h,elv-landscapefeat-h,elv-amenities-h,elv-barriers-h,elv-sports-h,elv-emergency-h,elv-accommodation-h,elv-restaurants-h,elv-shops-h,elv-tourism-h,elv-pubtrans-h,elv-car-h,elv-buildings-h,elv-borders-h,elv-background-h}]
2022-03-14 22:29:00.881 main@MapsforgeConfig:358 - Preferred map language : defined [FR]
2022-03-14 22:29:00.882 main@MapsforgeConfig:336 - Contrast stretch       : defined [0]
2022-03-14 22:29:00.883 main@MapsforgeConfig:336 - Gamma correction       : defined [1.0]
2022-03-14 22:29:00.888 main@MapsforgeConfig:502 - HillShading algorithm  : defined [simple(0.0,1.0)]
2022-03-14 22:29:00.889 main@MapsforgeConfig:336 - Hillshading magnitude  : defined [1.0]
2022-03-14 22:29:00.898 main@MapsforgeConfig:388 - DEM folder             : defined [F:\maps\dem\dtm1]
2022-03-14 22:29:00.898 main@MapsforgeConfig:358 - Renderer algorithm     : defined [direct]
2022-03-14 22:29:00.898 main@MapsforgeConfig:336 - Device scale factor    : defined [1.0]
2022-03-14 22:29:00.900 main@MapsforgeConfig:336 - User scale factor      : defined [1.0]
2022-03-14 22:29:00.900 main@MapsforgeConfig:336 - Text scale factor      : defined [1.0]
2022-03-14 22:29:00.901 main@MapsforgeConfig:336 - Symbol scale factor    : defined [1.0]
2022-03-14 22:29:00.901 main@MapsforgeConfig:336 - Browser cache ttl      : defined [2592000]
2022-03-14 22:29:00.901 main@MapsforgeConfig:358 - Out of range TMS url   : defined [https://a.tile.openstreetmap.fr/osmfr/{z}/{x}/{y}.png]
2022-03-14 22:29:00.902 main@MapsforgeConfig:336 - Server max queue size  : defined [128]
2022-03-14 22:29:00.902 main@MapsforgeConfig:336 - Server max thread(s)   : defined [32]
2022-03-14 22:29:00.903 main@MapsforgeConfig:336 - Server min thread(s)   : defined [0]
2022-03-14 22:29:00.903 main@MapsforgeConfig:336 - Connection idle timeout: defined [120000]
2022-03-14 22:29:00.904 main@MapsforgeConfig:415 - Server connector(s)    : defined [{h2c}]
```

```
2022-03-14 22:29:27.742 queue-45@MapsforgeHandler:514 - out-of-range redirect 'https://a.tile.openstreetmap.fr/osmfr/7/68/49.png'
2022-03-14 22:29:27.742 queue-59@MapsforgeHandler:514 - out-of-range redirect 'https://a.tile.openstreetmap.fr/osmfr/7/63/41.png'
2022-03-14 22:29:27.746 queue-53@MapsforgeHandler:526 - /7/66/43.png?textScale=1.2&userScale=0.8 [ms:438;idle:9;qs:0]
2022-03-14 22:29:28.729 queue-43@MapsforgeHandler:526 - /7/66/42.png?textScale=1.2&userScale=0.8 [ms:1564;idle:13;qs:0]
2022-03-14 22:29:28.854 queue-52@MapsforgeHandler:514 - out-of-range redirect 'https://a.tile.openstreetmap.fr/osmfr/7/67/44.png'
2022-03-14 22:29:28.854 queue-48@MapsforgeHandler:514 - out-of-range redirect 'https://a.tile.openstreetmap.fr/osmfr/7/60/42.png'
```

![image](https://user-images.githubusercontent.com/2869665/158265114-55d9b04c-1ff6-48ab-a959-4859671ba3fd.png)

**Comment by JFritzle** [comment]
Regarding resulting map I doubt that others will appreciate that feature as long as out-of-range tiles don't connect seamlessly to _useful_ content of offline map tiles.

**Comment by telemaxx** [comment]
No error in browser, just a white square.

Console show a null at the end:

![grafik](https://user-images.githubusercontent.com/22706487/158360549-763b0c81-9182-4972-8e01-8517d291e3c3.png)

**Comment by telemaxx** [comment]
> Regarding resulting map I doubt that others will appreciate that feature as long as out-of-range tiles don't connect seamlessly to _useful_ content of offline map tiles.

seamlessly would be better, but for an overview ans orientation it is usefull to me.

**Comment by telemaxx** [comment]
at least the white square is 256*256px

**Comment by nono303** [comment]
> No error in browser, just a white square.
> 
> Console show a null at the end:
> 
> ![grafik](https://user-images.githubusercontent.com/22706487/158360549-763b0c81-9182-4972-8e01-8517d291e3c3.png)

Is status code in browser 200, 302, 404 ?
According to https://github.com/telemaxx/mapsforgesrv/blob/master/mapsforgesrv/src/main/java/com/telemaxx/mapsforgesrv/MapsforgeHandler.java#L510 it'll explain in which case you are

Can you test again with request detail? (https://github.com/telemaxx/mapsforgesrv/blob/master/mapsforgesrv/src/main/java/com/telemaxx/mapsforgesrv/MapsforgeConfig.java#L101 to true)

**Comment by JFritzle** [comment]
_Console show a null at the end_

Method _logRequest_ calls _request.getQueryString()_.  If html request _http://127.0.0.1/{zoom}/{x}/{y}.png_ ends after _png_ not containing additional query data, _request.getQueryString()_ returns _null_ which gets appended to logged request. This behaviour was introduced with new logging method. Therefore as work around I am appending at least 1 question mark to html requests. Html request _http://127.0.0.1/{zoom}/{x}/{y}.png?_ gets logged without _null_.

**Comment by JFritzle** [comment]
To circumvent logging _null_, 
https://github.com/telemaxx/mapsforgesrv/blob/master/mapsforgesrv/src/main/java/com/telemaxx/mapsforgesrv/MapsforgeHandler.java#L324  
better should be replaced by
```
String msg = request.getPathInfo(); 
String query = request.getQueryString(); 
if (query != null) msg +=  "?" + query;
```

**Comment by nono303** [comment]
> Regarding resulting map I doubt that others will appreciate that feature as long as out-of-range tiles don't connect seamlessly to _useful_ content of offline map tiles.

It clearly depends on map range definition vs border content fit. 
ex, somewhere in Sardinia Z16, it looks great ;)
![image](https://user-images.githubusercontent.com/2869665/158389054-31b7157a-9858-4e77-848d-256c3c91e441.png)

**Comment by nono303** [comment]
...Can't understand why you don't trigger "out of range" with the code https://github.com/telemaxx/mapsforgesrv/blob/master/mapsforgesrv/src/main/java/com/telemaxx/mapsforgesrv/MapsforgeHandler.java#L509 if `tileBitmap == null` with your maps but **YES** @JFritzle ,embedding [world.map](https://download.mapsforge.org/maps/world/world.map) would be a better solution :)

**Comment by JFritzle** [comment]
**About  "out of range"**

I added some control statements:
- statement `logger.info("job_done");` before `if (tileBitmap != null) {` block
- statement `logger.info("outOfRange");` between `} else {` and `if(this.outOfRangeTms != null)`
- statement `logger.info("image width="+imageWidth+" height="+imageHeight);` after `int imageHeight = image.getHeight();`

My observation:
Wherever I locate the map window and whether given map(s) at that location covers corresponding tile or not, same number of  lines `job_done` and `image width=256 height=256` get logged. Line `outOfRange` gets never logged. Seems renderer job never returns value `null` for` tileBitmap` but fills tile with some background color if no data exist.

I wonder about different behaviour on my site and on yours. Might it depend on JRE or JDK version?

**About appending embedded world.map**

For a simple test add these lines after https://github.com/telemaxx/mapsforgesrv/blob/master/mapsforgesrv/src/main/java/com/telemaxx/mapsforgesrv/MapsforgeHandler.java#L190:
```
			InputStream inputStream = getClass().getResourceAsStream("/assets/mapsforgesrv/world.map");
			FileSystem fileSystem = MemoryFileSystemBuilder.newEmpty().build();
			Path rootPath = fileSystem.getPath("");
			Path worldMap = rootPath.resolve("world.map");
			Files.copy(inputStream, worldMap, StandardCopyOption.REPLACE_EXISTING);
			FileChannel mapFileChannel = FileChannel.open(worldMap, StandardOpenOption.READ);
			MapFile map = new MapFile(mapFileChannel);
			multiMapDataStore.addMapDataStore(map, true, true);
```
First, I know code could be easily optimized because exactly same lines already exist above.  
Second, appending world.map or not should be configurable by an additional parameter switch.

**Comment by telemaxx** [comment]
> > No error in browser, just a white square.
> > Console show a null at the end:
> > ![grafik](https://user-images.githubusercontent.com/22706487/158360549-763b0c81-9182-4972-8e01-8517d291e3c3.png)
> 
> Is status code in browser 200, 302, 404 ? According to https://github.com/telemaxx/mapsforgesrv/blob/master/mapsforgesrv/src/main/java/com/telemaxx/mapsforgesrv/MapsforgeHandler.java#L510 it'll explain in which case you are
> 
> Can you test again with request detail? (https://github.com/telemaxx/mapsforgesrv/blob/master/mapsforgesrv/src/main/java/com/telemaxx/mapsforgesrv/MapsforgeConfig.java#L101 to true)

public boolean LOGREQDET = true;

![grafik](https://user-images.githubusercontent.com/22706487/159242254-770138d6-beb2-456f-a1c5-58740e0911dd.png)


![grafik](https://user-images.githubusercontent.com/22706487/159242319-f5c33b48-bc7a-4dd5-be31-2277d43f20b4.png)

**Comment by telemaxx** [comment]
i think using the embedded [world.map] would be a good solution.

**Comment by JFritzle** [comment]
_using the embedded [world.map] would be a good solution_

What about (at least on my site) not working out-of-range map implementation? Remove from master code or keep when adding new world.map feature?

**Comment by telemaxx** [comment]
for now, I think better removing in master. 

@nono303 : what do you think?

**Comment by telemaxx** [comment]
OK, we keep it.

**Comment by telemaxx** [comment]
null is gone ;-)

![grafik](https://user-images.githubusercontent.com/22706487/159277288-8cdc0874-2c6f-48c2-bceb-8d3e0ce9a8b8.png)

**Comment by nono303** [comment]
@telemaxx before tagging 0.17.3, here is a quick patch without any feature changes that fix typo, upgrade deps and improve httpStatusResponseCode usage
[0.17.3.patch](https://github.com/telemaxx/mapsforgesrv/files/8316415/0.17.3.patch.txt)

**Comment by nono303** [comment]
:star2: I just find why it's not working for both of you: **renderer**
I use `direct` as you're configured with `database`
`direct` return a **null** tileBitmap where `database` return a **blank non-null** tileBitmap

So to provide an appropriated patch, do you think it's better to:
- disable out-of-range if renderer database is selected
- force direct renderer if out-of-range is configured

**Comment by telemaxx** [comment]
but seem so that Mytourbook does not like the redirect:

![grafik](https://user-images.githubusercontent.com/22706487/159310890-056a0dda-6573-4d1c-97d1-fc1780ec9093.png)


But i think this is a mytourbook problem.

**Comment by telemaxx** [comment]
i think we can close now this issue #29

**Comment by nono303** [comment]
I Had a look on mapsforge src and all can be explained:
https://github.com/mapsforge/mapsforge/issues/1299

**Comment by telemaxx** [comment]
hopefully the error fix is returning null, not background image...

**Comment by JFritzle** [comment]
While still working on appending embedded world map, I will also implement following workaround.

Solution which is already tested and prooved to be good:
```
			Tile tile = new Tile(x, y, (byte) z, requestedTileRenderSize);
			if (multiMapDataStore.supportsTile(tile)) {
				tileBitmap  = <result of executed {database,direct} render job>;
			} else {
				tileBitmap = null;
			}
```

This will make renderer dependent handling of `outofrange_tms` parameter unnecessary.

**Comment by nono303** [comment]
Many thx @JFritzle!
`multiMapDataStore.supportsTile` works great with database rendering and `outofrange_tms` configured
the corresponding 2 lines patch [database_oor.patch](https://github.com/telemaxx/mapsforgesrv/files/8321733/database_oor.patch.txt)


> i think we can close now this issue #29

Yes!

**Comment by JFritzle** [comment]
Meanwhile I have finished my enhancements and committed as version 0.17.4 to [my Java8 branch](https://github.com/JFritzle/mapsforgesrv). This includes optionally appending built-in world map as well as switching to outofrange_tms for database renderer too. Before creating a pull request _**please test**_ my repo's jar file which should work with higher Java versions too.

Case 1 - appending built-in world map by optional parameter `-wm`
![grafik](https://user-images.githubusercontent.com/62614244/159566584-403dea72-4c69-4999-856d-cf8f75bb840e.png)

Case 2 - redirecting out-of-range tiles by optional parameter `-tms https://a.tile.openstreetmap.fr/osmfr/{z}/{x}/{y}.png`
![grafik](https://user-images.githubusercontent.com/62614244/159566947-e4d024e4-62a9-44e1-b2f5-af8892bf70a8.png)
I notice `out-of-range redirect` messages, but unfortunately I do not see any tiles rendered by redirect server, neither with `database` nor with `direct` renderer. See output log:
```
Starte Mapsforge Tile Server (SRV) ...
java -Xmx1G -Xms256M -Xmn256M -Dsun.java2d.opengl=true -Dsun.java2d.renderer.log=true -Dsun.java2d.renderer.useLogger=true -Dsun.java2d.renderer.profile=speed -Dsun.java2d.renderer.useRef=hard -Dsun.java2d.renderer.useThreadLocal=true -Dsun.java2d.renderer.pixelWidth=2048 -Dsun.java2d.renderer.pixelHeight=2048 -Dsun.java2d.renderer.tileSize_log2=8 -Dsun.java2d.renderer.tileWidth_log2=8 -Dsun.java2d.renderer.subPixel_log2_X=2 -Dsun.java2d.renderer.subPixel_log2_Y=2 -Dsun.java2d.renderer.useFastMath=true -Dsun.java2d.render.bufferSize=524288 -jar D:/Programme/eclipse-home/j/git/mapsforgesrv/mapsforgesrv/bin/jars_ready2use/mapsforgesrv4java8.jar -if localhost -p 60815 -m D:/Landkarten/Mapsforge/maps/OpenAndroMaps/Dänemark.map -tms https://a.tile.openstreetmap.fr/osmfr/"z"/"x"/"y".png -t "C:/Program Files/MapsforgeSrv/themes/Elevate/Elevate.xml" -s elv-hiking -o elv-h_routes,elv-waymarks,elv-settlements-h,elv-landscapefeat-h,elv-amenities-h,elv-barriers-h,elv-sports-h,elv-accommodation-h,elv-restaurants-h,elv-tourism-h,elv-pubtrans-h,elv-car-h,elv-borders-h,elv-background-h -l de -r database -gc 0.90 -cs 0 -sft 1.25 -sfs 1.00 -sfu 1.00
Prozess 17836 (java) gestartet
(SRV) 2022-03-22 20:55:05.244 MapsforgeSrv - a mapsforge tile server. version: 0.17.4
(SRV) 2022-03-22 20:55:05.256 ################## CONFIG INIT ##################
(SRV) 2022-03-22 20:55:05.261 Listening TCP port        : defined [60815]
(SRV) 2022-03-22 20:55:05.261 Server interface          : defined [localhost]
(SRV) 2022-03-22 20:55:05.329 Map file(s)               : defined [{D:\Landkarten\Mapsforge\maps\OpenAndroMaps\Dänemark.map}]
(SRV) 2022-03-22 20:55:05.329 Theme file                : defined [C:\Program Files\MapsforgeSrv\themes\Elevate\Elevate.xml]
(SRV) 2022-03-22 20:55:05.329 Theme style               : defined [elv-hiking]
(SRV) 2022-03-22 20:55:05.350 Theme overlay(s)          : defined [{elv-h_routes,elv-waymarks,elv-settlements-h,elv-landscapefeat-h,elv-amenities-h,elv-barriers-h,elv-sports-h,elv-accommodation-h,elv-restaurants-h,elv-tourism-h,elv-pubtrans-h,elv-car-h,elv-borders-h,elv-background-h}]
(SRV) 2022-03-22 20:55:05.351 Preferred map language    : defined [de]
(SRV) 2022-03-22 20:55:05.351 Contrast stretch          : defined [0]
(SRV) 2022-03-22 20:55:05.353 Gamma correction          : defined [0.9]
(SRV) 2022-03-22 20:55:05.354 Hillshading magnitude     : default [1.0]
(SRV) 2022-03-22 20:55:05.354 DEM                       : default [undefined]
(SRV) 2022-03-22 20:55:05.354 Renderer algorithm        : defined [database]
(SRV) 2022-03-22 20:55:05.354 Device scale factor       : default [1.0]
(SRV) 2022-03-22 20:55:05.354 User scale factor         : defined [1.0]
(SRV) 2022-03-22 20:55:05.354 Text scale factor         : defined [1.25]
(SRV) 2022-03-22 20:55:05.354 Symbol scale factor       : defined [1.0]
(SRV) 2022-03-22 20:55:05.355 Browser cache ttl         : default [0]
(SRV) 2022-03-22 20:55:05.355 Out of range TMS url      : defined [https://a.tile.openstreetmap.fr/osmfr/{z}/{x}/{y}.png]
(SRV) 2022-03-22 20:55:05.355 Append built-in world map : default [undefined]
(SRV) 2022-03-22 20:55:05.355 Server max queue size     : default [256]
(SRV) 2022-03-22 20:55:05.355 Server max thread(s)      : default [8]
(SRV) 2022-03-22 20:55:05.355 Server min thread(s)      : default [0]
(SRV) 2022-03-22 20:55:05.355 Connection idle timeout   : default [0]
(SRV) 2022-03-22 20:55:05.355 Server connector(s)       : default [{http11}]
(SRV) 2022-03-22 20:55:05.368 Logging initialized @337ms to org.eclipse.jetty.util.log.Slf4jLog
(SRV) 2022-03-22 20:55:05.674 ################### MAPS INFO ###################
(SRV) 2022-03-22 20:55:05.690 'D:\Landkarten\Mapsforge\maps\OpenAndroMaps\Dänemark.map' supported languages: {de,en,da}
(SRV) 2022-03-22 20:55:05.827 ################ THEME FILE INFO ################
(SRV) 2022-03-22 20:55:05.828 Default Style: elv-hiking
(SRV) 2022-03-22 20:55:05.828 Stylename to use for '-s' option: 'elv-hiking' --> Wandern
(SRV) 2022-03-22 20:55:05.828 Stylename to use for '-s' option: 'elv-city' --> Stadt
(SRV) 2022-03-22 20:55:05.828 Stylename to use for '-s' option: 'elv-cycling' --> Radfahren
(SRV) 2022-03-22 20:55:05.828 Stylename to use for '-s' option: 'elv-mtb' --> Mountainbike
(SRV) 2022-03-22 20:55:05.828 The defined style 'elv-hiking' found
(SRV) 2022-03-22 20:55:05.904 ################# OVERLAY  INFO #################
(SRV) 2022-03-22 20:55:05.905 'elv-h_routes' enabled: true, title: '[R] Wanderrouten'
(SRV) 2022-03-22 20:55:05.905 'elv-h_s_routes' enabled: false, title: '[R] Betonte Routen'
(SRV) 2022-03-22 20:55:05.905 'elv-waymarks' enabled: true, title: '[R] Wegzeichen'
(SRV) 2022-03-22 20:55:05.906 'elv-road_surfaces-h' enabled: false, title: '[W] Erweiterte Straßenoberflächen'
(SRV) 2022-03-22 20:55:05.906 'elv-acc_allowed-h' enabled: false, title: '[W] Markierung erlaubter Fußwege'
(SRV) 2022-03-22 20:55:05.906 'elv-settlements-h' enabled: true, title: '[P] Ortsnamen'
(SRV) 2022-03-22 20:55:05.906 'elv-landscapefeat-h' enabled: true, title: '[P] Landschaftsmerkmale'
(SRV) 2022-03-22 20:55:05.906 'elv-amenities-h' enabled: true, title: '[P] Einrichtungen'
(SRV) 2022-03-22 20:55:05.906 'elv-barriers-h' enabled: true, title: '[P] Barrieren'
(SRV) 2022-03-22 20:55:05.906 'elv-sports-h' enabled: true, title: '[P] Sport'
(SRV) 2022-03-22 20:55:05.906 'elv-emergency-h' enabled: false, title: '[P] Notfall'
(SRV) 2022-03-22 20:55:05.906 'elv-accommodation-h' enabled: true, title: '[P] Unterkünfte'
(SRV) 2022-03-22 20:55:05.906 'elv-restaurants-h' enabled: true, title: '[P] Restaurants + Bars'
(SRV) 2022-03-22 20:55:05.906 'elv-shops-h' enabled: false, title: '[P] Geschäfte + Dienstleistungen'
(SRV) 2022-03-22 20:55:05.906 'elv-tourism-h' enabled: true, title: '[P] Tourismus + Kultur'
(SRV) 2022-03-22 20:55:05.906 'elv-pubtrans-h' enabled: true, title: '[P] Öffentlicher Verkehr'
(SRV) 2022-03-22 20:55:05.906 'elv-car-h' enabled: true, title: '[P] Auto'
(SRV) 2022-03-22 20:55:05.907 'elv-buildings-h' enabled: false, title: '[P] Besondere Gebäude'
(SRV) 2022-03-22 20:55:05.913 'elv-borders-h' enabled: true, title: '[A] Grenzen + besondere Gebiete'
(SRV) 2022-03-22 20:55:05.914 'elv-background-h' enabled: true, title: '[A] Kartenhintergrund'
(SRV) 2022-03-22 20:55:06.069 ################ STARTING SERVER ################
(SRV) 2022-03-22 20:55:06.073 jetty-9.4.45.v20220203; built: 2022-02-03T09:14:34.105Z; git: 4a0c91c0be53805e3fcffdcdcc9587d5301863db; jvm 1.8.0_321-b07
(SRV) 2022-03-22 20:55:06.142 Started ServerConnector@26f67b76{HTTP/1.1, (http/1.1, [proxy])}{127.0.0.1:60815}
(SRV) 2022-03-22 20:55:06.142 Started @1112ms
(SRV) 2022-03-22 20:55:06.142 server connector configured with accept queue size '128', idle timeout '0'
(SRV) 2022-03-22 20:55:06.142 job executor configured with threads min '0', max '8', idle timeout '0'
(SRV) 2022-03-22 20:55:06.142 job queue configured with max size '256'
(SRV) 2022-03-22 20:55:06.142 server is up and listening on 'localhost:60815'
(SRV) 2022-03-22 20:55:07.254 /0/0/0.png
Starte QMapShack (QMS) ...
"C:/Program Files/QMapShack/qmapshack.exe" --no-splash --style fusion
Prozess 6952 (qmapshack.exe) gestartet
(SRV) 2022-03-22 20:55:09.018 out-of-range redirect 'https://a.tile.openstreetmap.fr/osmfr/6/32/18.png'
(SRV) 2022-03-22 20:55:09.018 out-of-range redirect 'https://a.tile.openstreetmap.fr/osmfr/6/33/18.png'
(SRV) 2022-03-22 20:55:09.018 out-of-range redirect 'https://a.tile.openstreetmap.fr/osmfr/6/34/18.png'
(SRV) 2022-03-22 20:55:09.023 out-of-range redirect 'https://a.tile.openstreetmap.fr/osmfr/6/35/18.png'
(SRV) 2022-03-22 20:55:09.025 out-of-range redirect 'https://a.tile.openstreetmap.fr/osmfr/6/32/19.png'
(SRV) 2022-03-22 20:55:09.037 out-of-range redirect 'https://a.tile.openstreetmap.fr/osmfr/6/35/19.png'
(SRV) 2022-03-22 20:55:09.086 out-of-range redirect 'https://a.tile.openstreetmap.fr/osmfr/6/32/20.png'
(SRV) 2022-03-22 20:55:09.135 out-of-range redirect 'https://a.tile.openstreetmap.fr/osmfr/6/35/20.png'
(SRV) 2022-03-22 20:55:09.147 out-of-range redirect 'https://a.tile.openstreetmap.fr/osmfr/6/32/21.png'
(SRV) 2022-03-22 20:55:09.155 out-of-range redirect 'https://a.tile.openstreetmap.fr/osmfr/6/33/21.png'
(SRV) 2022-03-22 20:55:09.161 out-of-range redirect 'https://a.tile.openstreetmap.fr/osmfr/6/34/21.png'
(SRV) 2022-03-22 20:55:09.166 out-of-range redirect 'https://a.tile.openstreetmap.fr/osmfr/6/35/21.png'
(SRV) 2022-03-22 20:55:09.859 /6/33/20.png
(SRV) 2022-03-22 20:55:09.881 /6/34/20.png
(SRV) 2022-03-22 20:55:09.927 /6/33/19.png
(SRV) 2022-03-22 20:55:10.060 /6/34/19.png
```
As @telemaxx already told _Mytourbook does not like the redirect_, QMapShack application does not complain about redirection but seems to ignore it. Do I have to set additional/different parameter values? Please note: my code modifications are based on recent master branch!

**Comment by telemaxx** [comment]
merged in release 0.17.4 for java 8 and java 11
